docs/units/type_hints.rst

```diff
@@ -0,0 +1,136 @@
+Unit-Aware Type Annotations
+***************************
+
+.. |Quantity| replace:: :class:`~astropy.units.Quantity`
+
+While Python is a dynamically typed language, there is also support for type hints (and even static typing) using the annotations syntax of `PEP 484 <https://www.python.org/dev/peps/pep-0484/>`_.
+For a detailed guide on type hints, function annotations, and other related syntax see the `Real Python Guide <https://realpython.com/python-type-checking/#type-aliases>`_. 
+Here we how Astropy |Quantity| can be used in type hints and annotations and also include metadata about the associated units.
+
+
+For this and all further examples we assume the following imports
+
+::
+
+   >>> import typing as T
+   >>> import astropy.units as u
+   >>> from astropy.units import Quantity
+   >>> from astropy.units._typing import Annotated
+
+
+.. _quantity_type_annotation:
+
+Quantity Type Annotation
+========================
+
+A |Quantity| can easily be used as a type annotation,
+
+::
+
+   >>> x: Quantity = 2 * u.km
+
+or as a function annotation.
+
+::
+
+   >>> def func(x: Quantity) -> Quantity:
+   ...     return x
+
+However, depending on the computer, this frequently throws an exception when compiling Sphinx docs. A safer method to annotate Quantities is by leveraging the :mod:`~typing` library to create a type variable (note the ``~``).
+
+::
+
+   >>> QuantityType = T.TypeVar("QuantityType", bound=Quantity)
+
+Anything annotated with ``QuantityType`` is expected to be a |Quantity| or subclass thereof.
+
+Repeating the above examples,
+
+::
+
+   >>> x: QuantityType = 2 * u.km
+
+or as a function annotation.
+
+::
+
+   >>> def func(x: QuantityType) -> QuantityType:
+   ...     return x
+
+
+Preserving Units
+^^^^^^^^^^^^^^^^
+
+While the above annotations are useful for annotating the **value**'s type, it does not inform us of the other most important attribute of a |Quantity|: the **unit**.
+
+Unit information may be included by the syntax ``Quantity[unit or "physical_type", other_meta]``. **The unit metadata must be first.**
```

Comment:
Again, no need to tell the unit should be first - you just wrote that. Could also give the example without `other_meta` and note that one could append other metadata if needed.

Reply:
✅